3_visualize/src/sf_utils_shift.R

```diff
@@ -0,0 +1,287 @@
+    PR = list(scale = 3.15, shift = c(-95,75), rotation_deg=20),
+    VI = list(scale = 4, shift = c(-115,110), rotation_deg=20),
+    GU = list(scale = 5, shift = c(725, -575), rotation_deg=-90),
+    AS = list(scale = 5, shift = c(675, -30), rotation_deg=-45),
+    MP = list(scale = 1.5, shift = c(815,-600), rotation_deg=0),
+    # If there is no match, default to doing nothing to the object
+    list(scale = 1, shift = c(0,0), rotation_deg=0)
+  )
+}
+
+#' @title create a single sf object for regions outside of CONUS
+#' @desciption generate an sf object containing the spatial data 
+#' for one or more regions in the `maps` package "world" database. 
+#' @param region_abbrs one or more character strings that match the
+#' abbreviations listed in the `abbr_xwalk`
+#' @param proj_str character string representing the projection
+#' @value a single `sf` object with one or more polygons for the 
+#' current list of abbreviations
+generate_single_oconus_sf <- function(region_abbrs, proj_str) {
+  
+  if("AK" %in% region_abbrs) {
+    # If we are plotting AK, use the low-res version (and by 
+    # function simplicity, do any regions grouped with AK as 
+    # the low-res version, too)
+    ak_sf <- maps::map("world", "USA:Alaska", fill=TRUE, plot=FALSE) %>%
+      sf::st_as_sf() %>% 
+      st_transform(proj_str) %>% 
+      mutate(ID = "AK")
+    
+    # Scrub AK out of the list now
+    region_abbrs <- region_abbrs[!grepl("AK", region_abbrs)]
+  } 
+  
+  # If there were more `region_abbrs` passed in beyond just AK
+  # then use the high-res spatial data for those
+  if(length(region_abbrs) > 0) {
+    # Note that every time you call this function, it will load the
+    # high-resolution spatial shapefile.
+    regions_sf <- st_read('2_process/out/nws_states.shp') %>% 
+      select(ID = STATE) %>% 
+      filter(ID %in% region_abbrs) %>% 
+      rename(geom = geometry) %>% 
+      st_transform(proj_str)
+  }
+  
+  if(exists("ak_sf") & exists("regions_sf")) {
+    return(bind_rows(ak_sf, regions_sf))
+  } else if(exists("ak_sf") & !exists("regions_sf")) {
+    return(ak_sf)
+  } else if(!exists("ak_sf") & exists("regions_sf")){
+    return(regions_sf)
+  }
+}
+
+#' @title create a single, shifted sf object for regions outside of CONUS
+#' @desciption generate an sf object for regions outside of CONUS that 
+#' will be visible in a map view where CONUS is the focus. Currently
+#' functions for Alaska (`AK`), Hawaii (`HI`), Puerto Rico (`PR`), the
+#' U.S. Virgin Islands (`VI`), Guam (`GU`), American Samoa (`AS`), and
+#' the Northern Marianas (`MP`).
+#' @param proj_str character string representing the projection
+#' @value a single `sf` object a polygon per region shifted and scaled 
+#' to be visible on a CONUS map of the same projection.
+build_oconus_sf <- function(proj_str) {
+  
+  # Might need to treat PR & VI as one
+  region_abbr_list <- list(
+    "AK",
+    "HI",
+    "PR", 
+    "VI", 
+    "GU",
+    "AS",
+    "MP"
+  )
```

Comment:
Relatedly, this function and the one above could take a list of territories to include

Reply:
agree! Would be easy to change that - I wrote all the helper functions to handle different combos of this list